docs/LANGUAGE.md

```diff
@@ -0,0 +1,767 @@
+
+The examples in this section make use of the data defined
+in the [Examples](#examples) section.
+
+### <a name="set-documents"></a> Generating Sets
+
+The following rule documents a set containing the hostnames of all servers:
+
+```opalog
+hostnames[] = name :- sites[].servers[].hostname = name
+```
+
+When we query for the content of "hostnames" we see the same data as we would if we queried using the `sites[].servers[].hostname` reference directly:
+
+```
+hostnames
+# [
+#   "hydrogen",
+#   "helium",
+#   "lithium",
+#   "berylium",
+#   "boron",
+#   "carbon",
+#   "nitrogen",
+#   "oxygen"
+# ]
+```
+
+This example introduces a few important aspects of Opalog.
+
+First, the rule defines a set document where the contents are defined by the
+variable "name". We know this rule defines a set document because the head
+does not include a key, i.e., all rules follow this syntax:
+
+```
+rule    ::= head ":-" body
+head    ::= name "[" [key] "]" [value]
+name    ::= var
+key     ::= var
+value   ::= var
+```
+
+> Set documents are collections of values without keys. OPA represents set documents as arrays when serializing to JSON or other formats which do not support a set data type. The important distinction between sets and arrays or objects is that sets are un-keyed while arrays and objects are keyed, i.e., you cannot refer to the index of an element within a set. While you cannot refer to sets as arrays or objects, you can refer to the values of arrays or objects as sets.
+
+Second, the `sites[].servers[].hostname` fragment selects the "hostname" attribute from all of the objects in the "servers" collection. From reading the fragment in isolation we cannot tell whether the fragment refers to sets, arrays, or objects. We only know that it refers to a collections of values.
+
+> Under the hood, OPA implicitly adds a variable to the reference for the purpose of evaluation, e.g., `sites[].servers[].hostname` becomes `sites[$a].servers[$b].hostname`. This translation happens for arrays, objects, AND sets. In the case of sets, OPA can determine when a reference points to a set and does not attempt to bind a value to it (i.e., an element index).
+
+Third, the `sites[].servers[].hostname = name` expression binds the value of the "hostname" attribute to the variable "name", which is also declared in the head of the rule.
+
+### <a name="object-documents"></a> Generating Objects
+
+Rules that define objects are very similar to rules that define sets.
+
+```opalog
+apps_by_hostname[hostname] = app :-
+    sites[].servers[] = server,
+    server.hostname = hostname,
+    apps[i].servers[] = server.name,
+    apps[i].name = app
+```
+
+The rule above defines an object that maps hostnames to app names. The main
+difference between this rule and one which defines a set is the rule head: in
+addition to declaring a value, the rule head also declares a key for the
+document.
+
+The result:
+
+```
+apps_by_hostname["helium"]
+# "web"
+```
```

Comment:
We should probably describe what happens when a single key is assigned 2 values.  Not sure what the right thing to do here is.  The top-down algorithm doesn't actually care since if you ask for all the y such that obj["abc"] = y, it will happily return multiple values.  Or if you ask if obj["abc"]="def" it will return true, and if you ask if obj["abc"] = "ghi" it will return true.  The only time multiple values for a single key will show up is when we need to construct a single dictionary of the key/value pairs to hand off to something else.